File: Autonomous Mobile Robots Course/Chapter7/Lesson4/Chapter7_Lesson4.m
% Chapter7_Lesson4.m
% Kalman-Filter Localization for AMR — Lesson 4
% EKF tuning via innovation statistics (NIS) for a unicycle + GPS example.
%
% This script is self-contained and uses only base MATLAB.
% Optional: build a simple Simulink shell (requires Simulink) by calling:
%   build_simulink_shell_Chapter7_Lesson4();

function Chapter7_Lesson4()
  dt = 0.1;
  [xTrue, uMeas, zGps] = simulate_data(60.0, dt, 2);

  % Initial guesses (deliberately imperfect)
  p.sigma_v = 0.03;
  p.sigma_w = 0.01;
  p.sigma_gps = 0.3;
  p.q_scale = 1.0;
  p.r_scale = 1.0;

  for it = 1:4
    [nisList, neesList] = run_filter_stats(xTrue, uMeas, zGps, dt, p);

    nisMean = mean(nisList);
    neesMean = mean(neesList);

    fprintf('Iter %d: r_scale=%.3f, q_scale=%.3f\\n', it, p.r_scale, p.q_scale);
    fprintf('  mean NIS=%.3f (target ~ 2), mean NEES=%.3f (target ~ 3)\\n', nisMean, neesMean);

    % Heuristic update for R using NIS
    gain = min(5.0, max(0.2, nisMean/2.0));
    p.r_scale = p.r_scale * gain;
  end

  disp('Final tuned parameters:');
  disp(p);
end

function a = wrap_angle(a)
  a = mod(a + pi, 2*pi) - pi;
end

function x2 = f_unicycle(x, u, dt)
  px = x(1); py = x(2); th = x(3);
  v = u(1);  w = u(2);
  x2 = zeros(3,1);
  x2(1) = px + v*dt*cos(th);
  x2(2) = py + v*dt*sin(th);
  x2(3) = wrap_angle(th + w*dt);
end

function F = jacobian_F(x, u, dt)
  th = x(3); v = u(1);
  F = eye(3);
  F(1,3) = -v*dt*sin(th);
  F(2,3) =  v*dt*cos(th);
end

function L = jacobian_L(x, dt)
  th = x(3);
  L = zeros(3,2);
  L(1,1) = dt*cos(th);
  L(2,1) = dt*sin(th);
  L(3,2) = dt;
end

function Q = make_Q(x, u, dt, p)
  L = jacobian_L(x, dt);
  M = diag([p.sigma_v^2, p.sigma_w^2]);
  Q = p.q_scale * (L*M*L');
  %#ok<NASGU>
end

function zhat = h_gps(x)
  zhat = x(1:2);
end

function H = jacobian_H_gps()
  H = zeros(2,3);
  H(1,1) = 1;
  H(2,2) = 1;
end

function R = make_R_gps(p)
  R = p.r_scale * diag([p.sigma_gps^2, p.sigma_gps^2]);
end

function [nisList, neesList] = run_filter_stats(xTrue, uMeas, zGps, dt, p)
  N = size(xTrue,1);
  x = [0;0;0];
  P = diag([1,1,(10*pi/180)^2]);

  H = jacobian_H_gps();
  nisList = zeros(N,1);
  neesList = zeros(N,1);

  for k = 1:N
    % Predict
    xPred = f_unicycle(x, uMeas(k,:)', dt);
    F = jacobian_F(x, uMeas(k,:)', dt);
    Q = make_Q(x, uMeas(k,:)', dt, p);
    PPred = F*P*F' + Q;

    % Update
    y = zGps(k,:)' - h_gps(xPred);
    S = H*PPred*H' + make_R_gps(p);
    K = PPred*H' / S;
    x = xPred + K*y;
    x(3) = wrap_angle(x(3));
    P = (eye(3) - K*H)*PPred;

    nisList(k) = y' / S * y;
    e = x - xTrue(k,:)';
    e(3) = wrap_angle(e(3));
    neesList(k) = e' / P * e;
  end
end

function [xTrue, uMeas, zGps] = simulate_data(T, dt, seed)
  rng(seed);
  N = floor(T/dt);
  xTrue = zeros(N,3);
  uTrue = zeros(N,2);

  for k = 1:N
    t = (k-1)*dt;
    uTrue(k,1) = 1.0 + 0.2*sin(0.2*t);
    uTrue(k,2) = 0.2*sin(0.1*t);
    if k > 1
      xTrue(k,:) = f_unicycle(xTrue(k-1,:)', uTrue(k-1,:)', dt)';
    end
  end

  sigma_v_meas = 0.08;
  sigma_w_meas = 0.04;
  sigma_gps_meas = 0.6;

  uMeas = uTrue + randn(N,2).* [sigma_v_meas, sigma_w_meas];
  zGps  = xTrue(:,1:2) + randn(N,2).* sigma_gps_meas;
end

function build_simulink_shell_Chapter7_Lesson4()
  % Creates a minimal Simulink shell illustrating where EKF prediction/update
  % blocks go. You still need to implement the EKF equations inside a
  % MATLAB Function block or use a suitable toolbox implementation.

  mdl = 'Chapter7_Lesson4_Simulink';
  if bdIsLoaded(mdl), close_system(mdl, 0); end
  new_system(mdl); open_system(mdl);

  add_block('simulink/Sources/From Workspace', [mdl '/uMeas']);
  add_block('simulink/Sources/From Workspace', [mdl '/zGps']);
  add_block('simulink/User-Defined Functions/MATLAB Function', [mdl '/EKF_Core']);
  add_block('simulink/Sinks/To Workspace', [mdl '/xEst']);

  set_param([mdl '/uMeas'], 'VariableName', 'uMeas');
  set_param([mdl '/zGps'], 'VariableName', 'zGps');
  set_param([mdl '/xEst'], 'VariableName', 'xEst');

  add_line(mdl, 'uMeas/1', 'EKF_Core/1');
  add_line(mdl, 'zGps/1', 'EKF_Core/2');
  add_line(mdl, 'EKF_Core/1', 'xEst/1');

  set_param(mdl, 'StopTime', '60');
  save_system(mdl);
  disp(['Created Simulink shell: ' mdl '.slx']);
end

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter7/Lesson4/Chapter7_Lesson4.cpp
/* 
Chapter7_Lesson4.cpp
Kalman-Filter Localization for AMR — Lesson 4
EKF tuning via innovation statistics (NIS) for a unicycle + GPS example.

Build (example):
  g++ -O2 -std=c++17 Chapter7_Lesson4.cpp -I /usr/include/eigen3 -o ekf_tuning

Dependencies:
  Eigen (header-only)
*/

#include <Eigen/Dense>
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using Vec2 = Eigen::Vector2d;
using Vec3 = Eigen::Vector3d;
using Mat2 = Eigen::Matrix2d;
using Mat3 = Eigen::Matrix3d;
using Mat32 = Eigen::Matrix<double,3,2>;
using Mat23 = Eigen::Matrix<double,2,3>;

static double wrap_angle(double a) {
    while (a > M_PI)  a -= 2.0*M_PI;
    while (a <= -M_PI) a += 2.0*M_PI;
    return a;
}

struct NoiseParams {
    double sigma_v;     // process noise on v
    double sigma_w;     // process noise on w
    double sigma_gps;   // measurement noise
    double q_scale = 1.0;
    double r_scale = 1.0;
};

static Vec3 f_unicycle(const Vec3& x, const Vec2& u, double dt) {
    const double px = x(0), py = x(1), th = x(2);
    const double v = u(0), w = u(1);
    Vec3 x2;
    x2(0) = px + v*dt*std::cos(th);
    x2(1) = py + v*dt*std::sin(th);
    x2(2) = wrap_angle(th + w*dt);
    return x2;
}

static Mat3 jacobian_F(const Vec3& x, const Vec2& u, double dt) {
    Mat3 F = Mat3::Identity();
    const double th = x(2);
    const double v = u(0);
    F(0,2) = -v*dt*std::sin(th);
    F(1,2) =  v*dt*std::cos(th);
    return F;
}

static Mat32 jacobian_L(const Vec3& x, double dt) {
    Mat32 L; L.setZero();
    const double th = x(2);
    L(0,0) = dt*std::cos(th);
    L(1,0) = dt*std::sin(th);
    L(2,1) = dt;
    return L;
}

static Mat3 make_Q(const Vec3& x, const Vec2& u, double dt, const NoiseParams& p) {
    Mat32 L = jacobian_L(x, dt);
    Eigen::Matrix2d M = Eigen::Matrix2d::Zero();
    M(0,0) = p.sigma_v*p.sigma_v;
    M(1,1) = p.sigma_w*p.sigma_w;
    Mat3 Q = p.q_scale * (L * M * L.transpose());
    (void)u; // u not used here (kept for symmetry with theory)
    return Q;
}

static Vec2 h_gps(const Vec3& x) {
    return Vec2(x(0), x(1));
}

static Mat23 jacobian_H() {
    Mat23 H; H.setZero();
    H(0,0) = 1.0;
    H(1,1) = 1.0;
    return H;
}

static Mat2 make_R(const NoiseParams& p) {
    Mat2 R = Mat2::Zero();
    R(0,0) = p.sigma_gps*p.sigma_gps;
    R(1,1) = p.sigma_gps*p.sigma_gps;
    return p.r_scale * R;
}

static double nis(const Vec2& innov, const Mat2& S) {
    return innov.transpose() * S.inverse() * innov;
}

struct SimData {
    std::vector<Vec3> x_true;
    std::vector<Vec2> u_meas;
    std::vector<Vec2> z_gps;
};

static SimData simulate(double T, double dt, unsigned seed) {
    const int N = static_cast<int>(T/dt);
    SimData d;
    d.x_true.resize(N);
    d.u_meas.resize(N);
    d.z_gps.resize(N);

    std::mt19937 gen(seed);
    std::normal_distribution<double> n01(0.0, 1.0);

    // True controls
    std::vector<Vec2> u_true(N);
    for (int k=0; k<N; ++k) {
        const double t = k*dt;
        const double v = 1.0 + 0.2*std::sin(0.2*t);
        const double w = 0.2*std::sin(0.1*t);
        u_true[k] = Vec2(v,w);
        if (k>0) d.x_true[k] = f_unicycle(d.x_true[k-1], u_true[k-1], dt);
    }

    const double sigma_v_meas = 0.08;
    const double sigma_w_meas = 0.04;
    const double sigma_gps_meas = 0.6;

    for (int k=0; k<N; ++k) {
        d.u_meas[k](0) = u_true[k](0) + sigma_v_meas*n01(gen);
        d.u_meas[k](1) = u_true[k](1) + sigma_w_meas*n01(gen);
        d.z_gps[k](0) = d.x_true[k](0) + sigma_gps_meas*n01(gen);
        d.z_gps[k](1) = d.x_true[k](1) + sigma_gps_meas*n01(gen);
    }
    return d;
}

static void run_once(const NoiseParams& p, const SimData& d, double dt,
                     std::vector<double>& nis_list) {
    const int N = static_cast<int>(d.x_true.size());
    Vec3 x(0.0,0.0,0.0);
    Mat3 P = Mat3::Zero();
    P(0,0)=1.0; P(1,1)=1.0; P(2,2)=std::pow(10.0*M_PI/180.0,2);

    Mat23 H = jacobian_H();

    nis_list.clear();
    nis_list.reserve(N);

    for (int k=0; k<N; ++k) {
        // Predict
        Vec3 x_pred = f_unicycle(x, d.u_meas[k], dt);
        Mat3 F = jacobian_F(x, d.u_meas[k], dt);
        Mat3 Q = make_Q(x, d.u_meas[k], dt, p);
        Mat3 P_pred = F*P*F.transpose() + Q;

        // Update (GPS)
        Vec2 y = d.z_gps[k] - h_gps(x_pred);
        Mat2 S = H*P_pred*H.transpose() + make_R(p);
        Eigen::Matrix<double,3,2> K = P_pred*H.transpose()*S.inverse();
        x = x_pred + K*y;
        x(2) = wrap_angle(x(2));
        P = (Mat3::Identity() - K*H)*P_pred;

        nis_list.push_back(nis(y,S));
    }
}

int main() {
    const double dt = 0.1;
    SimData d = simulate(60.0, dt, 2);

    NoiseParams p;
    p.sigma_v = 0.03;
    p.sigma_w = 0.01;
    p.sigma_gps = 0.3;

    std::vector<double> nis_list;

    for (int it=0; it<4; ++it) {
        run_once(p, d, dt, nis_list);
        double mean = 0.0;
        for (double v: nis_list) mean += v;
        mean /= static_cast<double>(nis_list.size());

        std::cout << "Iter " << it << ": r_scale=" << p.r_scale
                  << ", mean NIS=" << mean << " (target ~ 2)" << std::endl;

        // heuristic update
        const double gain = std::min(5.0, std::max(0.2, mean/2.0));
        p.r_scale *= gain;
    }

    std::cout << "Final r_scale=" << p.r_scale << std::endl;
    return 0;
}
